### PR TITLE
Feature/acs 8005 jdbc golden merging

### PR DESCRIPTION
**Pull request by fingoltin**

### CODE REVIEW

File: pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java [hunk 1]
```diff
@@ -97,7 +100,7 @@ public class ConvertResourceFn extends FetchSearchPageFn<HapiRowDescriptor> {
 
   public void writeResource(HapiRowDescriptor element)
       throws IOException, ParseException, SQLException, ViewApplicationException, ProfileException {
-    String resourceId = element.resourceId();
+    String fhirId = element.fhirId();
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java:104]: **question:** I wonder how they'll see this change. May be "breaking" for them.

File: pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java [hunk 2]
```diff
@@ -192,6 +200,40 @@ private void setMetaTags(HapiRowDescriptor element, Meta meta) {
     }
   }
 
+  // TODO: Test StringUtils.replaceEach performance
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java]: **todo:** Should probably do this?

File: pipelines/batch/src/main/java/com/google/fhir/analytics/FhirSearchUtil.java [hunk 3]
```diff
@@ -353,7 +353,7 @@ public String getGoldenResourceFromId(String fhirId, boolean treatPossibleMatche
     Parameters response = null;
     try {
       response =
-         fetchUtil
+          fetchUtil
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/FhirSearchUtil.java]: **suggestion:** revert?

File: pipelines/batch/src/main/java/com/google/fhir/analytics/MdmLink.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+public class MdmLink implements Serializable {
+
+  private static final long serialVersionUID = 1L;
+
+  // the resource that contains references to golden resources, e.g. Encounter
+  String resourceId;
+
+  // the resource that is the golden resource, e.g. Patient
+  String sourceFhirId;
+  String goldenFhirId;
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/MdmLink.java:36]: **question:** Should these be `private`?

File: pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java (hunk 1 above)
Comment (fingoltin) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java:104]: Yeah, might need a "compatibility" option but it didn't really make sense

File: pipelines/batch/src/main/java/com/google/fhir/analytics/FhirSearchUtil.java (hunk 3 above)
Comment (fingoltin) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/FhirSearchUtil.java]: That was probably just the formatter

File: pipelines/batch/src/main/java/com/google/fhir/analytics/MdmLink.java (hunk 4 above)
Comment (fingoltin) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/MdmLink.java:36]: The other "data" classes' fields aren't private, but I don't know if there's a reason

File: pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcFetchHapi.java [hunk 5]
```diff
@@ -172,62 +218,108 @@ public FetchRowsJdbcIo(
...
+
+      builder =
+          new StringBuilder(
+              "SELECT res_link.src_resource_id, golden.fhir_id AS golden_fhir_id, source.fhir_id AS"
+                  + " source_fhir_id FROM hfj_res_link res_link JOIN mpi_link mdm_link ON"
+                  + " res_link.target_resource_id = mdm_link.target_pid JOIN hfj_resource golden ON"
+                  + " golden.res_id = mdm_link.golden_resource_pid JOIN hfj_resource source ON"
+                  + " source.res_id = mdm_link.target_pid WHERE res_link.source_resource_type = ?"
+                  + " AND res_link.src_resource_id % ? = ? AND"
+                  + " mdm_link.match_result = 2 OR mdm_link.match_result = 1");
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcFetchHapi.java]: Should we just make this "or" added based on the property? Or did you remove that setting?

Comment (fingoltin) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcFetchHapi.java]: I'm thinking we do that in a separate PR as it complicates the other `SELECT` quite a bit.

File: pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java (hunk 1 above)
Comment (michealharrington) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java:104]: Yeah, we can wait for them to complain on the PR to their code base

File: pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java [hunk 6]
```diff
@@ -115,7 +115,9 @@ abstract class FetchSearchPageFn<T> extends DoFn<T, KV<String, Integer>> {
 
   private final int maxPoolSize;
 
-  private static final Cache<String, String> goldenResources = Caffeine.newBuilder().maximumSize(10000).build();;
+  private static final Cache<String, String> goldenResources =
+      Caffeine.newBuilder().maximumSize(10000).build();
+  ;
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java]: 2 `;;`?

Comment (fingoltin) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java]: Are we good to remove fetch mode support?

Comment (michealharrington) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java]: It does not scale, and we should remove the code we don't want to maintain.

File: pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtlOptions.java [hunk 7]
```diff
@@ -265,4 +265,18 @@ public interface FhirEtlOptions extends BasePipelineOptions {
...
+  @Description(
+      "Whether to use HAPI-FHIR's MDM service to replace Resource references with their Golden"
+          + " Resource counterparts")
+  @Default.Boolean(false)
+  Boolean getMapToGoldenResources();
+
+  void setMapToGoldenResources(Boolean mapToGoldenResources);
+
+  @Description("Comma separated list of resources to treat as MDM Resources.")
+  @Default.String("Patient")
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtlOptions.java]: **todo:** I don't think this should have a default value. From the point of view of getting this merged into the mainline, not all users of `fhir-data-pipes` will be using HAPI-FHIR's MDM.

File: pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcFetchHapi.java [hunk 8]
```diff
@@ -135,34 +137,78 @@ public ResourceTag mapRow(ResultSet resultSet) throws Exception {
...
 
     private final String query;
 
     private final String tagQuery;
 
+    private final String mdmQuery;
+
+    private final boolean isMdmResource;
+    private final boolean mapToGoldenResources;
+    private String identifiersQuery;
```
Comment (michealharrington) [inline pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcFetchHapi.java]: **question:** All the rest of the fields in the class are `final`, should this one be as well?

File: pipelines/controller/src/main/java/com/google/fhir/analytics/DataProperties.java [hunk 9]
```diff
@@ -109,6 +109,12 @@ public class DataProperties {
 
   public String sinkUserName;
 
+  private boolean mapToGoldenResources;
+
+  private String mdmResourceList;
+
+  private String structureDefinitionsDir;
```
Comment (michealharrington) [inline pipelines/controller/src/main/java/com/google/fhir/analytics/DataProperties.java]: **question:** What does this have to do with your MDM work?

**Review by michealharrington — CHANGES_REQUESTED** [review]

File: pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtlOptions.java (hunk 7 above)
Comment (fingoltin) [reply pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtlOptions.java]: Good catch, I had reverted it to a blank string but it got lost in one of my merges.

File: pipelines/controller/src/main/java/com/google/fhir/analytics/DataProperties.java (hunk 9 above)
Comment (fingoltin) [reply pipelines/controller/src/main/java/com/google/fhir/analytics/DataProperties.java]: Nada

**Review by michealharrington — APPROVED** [review]

**Comment by fingoltin** [comment]
Not sure how that test ever passed.

**Review by michealharrington — APPROVED** [review]
